### ISSUE TITLE
optimize softmax computation

### ISSUE THREAD

**Issue by Jianqoq** [Jianqoq/Hpt#16]
currently softmax is implmented but performance is not good, implement softmax based on the `reduce` kernel and existing softmax computation logic. see `tensor_dyn\src\ops\cpu\common_reduce.rs` as a reference, implement softmax based on the reduction logic.
we should use max as softmax's body, add additional computation based on that.

first implement non simd version, then start to compare performance with libtorch and implment simd kernels similar with reduce kernels, see `tensor_dyn\src\ops\cpu\kernels\reduce_kernels.rs`

- [ ] softmax is doing correct computation comparing with libtorch without vectorization
- [ ] look at pytorch documentation, let our softmax has same behaviour as pytorch, 
- [ ] vectorize softmax for `avx2`
- [ ] vectorize softmax for `avx512f`
- [ ] vectorize softmax for `neon` and `sse`
- [ ] write test case for gather 
- [ ] test pass for contiguous `tensor`
- [ ] test pass for uncontiguous `tensor`
- [ ] test pass for sub contiguous `tensor`
- [ ] test pass for sub uncontiguous `tensor`
